source/communication_interfaces/src/sockets/UDPSocket.cpp

```diff
@@ -0,0 +1,135 @@
+  this->set_parameters(parameters);
+}
+
+UDPSocket::~UDPSocket() {
+  UDPSocket::close();
+}
+
+void UDPSocket::open_socket(bool bind_socket) {
+  if (this->buffer_size_->is_empty()) {
+    throw exceptions::SocketConfigurationException("Parameter 'buffer_size' is empty.");
+  }
+
+  try {
+    this->addr_len_ = sizeof(this->server_address_);
+    this->server_address_.sin_family = AF_INET;
+    this->server_address_.sin_addr.s_addr = inet_addr(this->ip_address_->get_value().c_str());
+    this->server_address_.sin_port = htons(this->port_->get_value());
+  } catch (const std::exception& ex) {
+    throw exceptions::SocketConfigurationException("Socket configuration failed: " + std::string(ex.what()));
+  }
+
+  this->server_fd_ = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+  if (this->server_fd_ < 0) {
+    throw exceptions::SocketConfigurationException("Opening socket failed");
+  }
+  if (this->enable_reuse_ && this->enable_reuse_->get_value()) {
+    const int opt_reuse = 1;
+    if (setsockopt(this->server_fd_, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt_reuse, sizeof(opt_reuse)) != 0) {
+      throw exceptions::SocketConfigurationException("Setting socket option (enable reuse) failed");
+    }
+  }
+  if (bind_socket) {
+    if (bind(this->server_fd_, (sockaddr*) &(this->server_address_), sizeof(server_address_)) != 0) {
+      throw exceptions::SocketConfigurationException("Binding socket failed.");
+    }
+  }
+
+  if (this->timeout_duration_sec_->get_value() > 0.0) {
+    timeval timeout{};
+    auto secs = std::floor(this->timeout_duration_sec_->get_value());
+    timeout.tv_sec = static_cast<long>(secs);
+    timeout.tv_usec = static_cast<long>((this->timeout_duration_sec_->get_value() - secs) * 1e6);
+    if (setsockopt(this->server_fd_, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout)) != 0) {
+      throw exceptions::SocketConfigurationException("Setting socket timeout failed.");
+    }
+  }
+}
+
+bool UDPSocket::recvfrom(const sockaddr_in& address, ByteArray& buffer) {
+  std::vector<char> local_buffer(this->buffer_size_->get_value());
+  auto receive_length = ::recvfrom(
+      this->server_fd_, local_buffer.data(), this->buffer_size_->get_value(), 0, (sockaddr*) &(address),
+      &(this->addr_len_));
+  if (receive_length < 0) {
+    return false;
+  }
+  local_buffer.at(receive_length) = 0;
+  buffer.reset();
+  buffer.load(local_buffer.data(), receive_length);
+  return true;
+}
+
+bool UDPSocket::sendto(const sockaddr_in& address, const ByteArray& buffer) const {
+  std::vector<char> local_buffer;
+  buffer.copy_to(local_buffer);
+  int send_length = ::sendto(
+      this->server_fd_, local_buffer.data(), local_buffer.size(), 0, (sockaddr*) &(address), this->addr_len_);
+  return send_length == static_cast<int>(local_buffer.size());
+}
+
+void UDPSocket::validate_and_set_parameter(const std::shared_ptr<ParameterInterface>& parameter) {
+  // TODO can we put that in ParameterMap?
```

Comment:
I agree, I think it can be added to `assert_parameter_valid`. In rare cases it might be considered a breaking change, but only at runtime in particular edge cases. Currently that function would throw `std::out_of_range` if the parameter isn't in the map, so it would only affect code that specifically tries to catch out of range exceptions (those would no longer be caught if instead an `InvalidParameterException` is returned). I think it's safer for it to be a single validation with a single exception type, and so I would agree to include it in the next minor release.

Reply:
Alright, I put it in the backlog